### PR TITLE
Cleanup local code transport execution

### PR DESCRIPTION
**Pull request by kimchy**
Now that we handle automatically the local execution within the transport service, we can remove parts of the code that handle it in actions.

### CODE REVIEW

File: src/main/java/org/elasticsearch/search/action/SearchServiceTransportAction.java [hunk 1]
```diff
@@ -430,134 +348,78 @@ public void sendExecuteFetchScroll(DiscoveryNode node, final ShardFetchRequest r
...
-                    listener.onFailure(exp);
-                }
-
-                @Override
-                public String executor() {
-                    return ThreadPool.Names.SAME;
-                }
-            });
-        }
+        transportService.sendRequest(node, SCAN_ACTION_NAME, request, new BaseTransportResponseHandler<QuerySearchResult>() {
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/search/action/SearchServiceTransportAction.java]: these handlers all look the same can we have a new abstract base class where we pass the handler and the threadpool as the ctor arg  and only override the newInstance method?

**Comment by s1monw** [comment]
left one comment otherwise LGTM

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 2]
```diff
@@ -384,6 +338,8 @@ protected void doStart() throws ElasticsearchException {
                 }
 
                 foundPrimary = true;
+                // we rely today on the primary shortcut to not get into infinite loop in chasing primaries
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: +1 to having a TransportPrimaryShardOperationAction that will chase the primary (if I get what you mean correctly)

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 3]
```diff
@@ -627,67 +583,26 @@ void performOnReplica(final ReplicationState state, final ShardRouting shard, fi
...
-                        shardOperationOnReplica(shardRequest);
-                        state.onReplicaSuccess();
-                    } catch (Throwable e) {
-                        failReplicaIfNeeded(shard.index(), shard.id(), e);
-                        state.onReplicaFailure(nodeId, e);
+                @Override
+                public void handleException(TransportException exp) {
+                    state.onReplicaFailure(nodeId, exp);
+                    logger.trace("[{}] Transport failure during replica request [{}] ", exp, node, internalRequest.request());
+                    if (!ignoreReplicaException(exp)) {
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: I think we need an extra check here to see if this was a local or remote execution. If local, we'll have double logging and double shard failed messages.

**Comment by bleskes** [comment]
Nice cleanup. Left one comment.

**Comment by kimchy** [comment]
@s1monw great idea, I went ahead and took the cleanup a bit more, tell me what you think...

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java (hunk 2 above)
Comment (kimchy) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: yea, but will leave it for another change, for now kept the comment toe explain

Comment (bleskes) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: Agreed. Totally a different change.

**Comment by kimchy** [comment]
@bleskes I actually reverted the replica part, its not simple to do this refactor now (like forced execution, ...), how does it look now?

**Pull request closed without merging**